addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/xml/DeviceDescriptorXML.java

```diff
@@ -0,0 +1,180 @@
+        public String toString() {
+            return commands.stream().collect(Collectors.joining(";"));
+        }
+    }
+
+    public class SystemDescriptor extends HasCommands {
+
+        public SystemDescriptor(Element element) {
+            super(element);
+        }
+    }
+
+    public class ZoneDescriptor extends HasCommands {
+
+        public final Zone zone;
+
+        public ZoneDescriptor(Zone zone, Element element) {
+            super(element);
+            this.zone = zone;
+            logger.trace("Zone {} has commands: {}", zone, super.toString());
+        }
+    }
+
+    public class FeatureDescriptor extends HasCommands {
+
+        public final Feature feature;
+
+        public FeatureDescriptor(Feature feature, Element element) {
+            super(element);
+            this.feature = feature;
+            logger.trace("Feature {} has commands: {}", feature, super.toString());
+        }
+    }
+
+    /**
+     * Get the descriptor XML from the AVR and parse
+     * @param con
+     */
+    public void load(XMLConnection con) {
+
+        // Get and store the Yamaha Description XML. This will be used to detect proper element naming in other areas.
+        Node descNode = tryGetDescriptor(con);
+
+        system = buildFeatureLookup(descNode, "Unit",
+                tag -> tag,
+                (tag, e) -> new SystemDescriptor(e))
+                .getOrDefault("System", null); // there will be only one System entry
+
+        zones = buildFeatureLookup(descNode, "Subunit",
+                tag -> YamahaUtils.tryParseEnum(Zone.class, tag),
+                (zone, e) -> new ZoneDescriptor(zone, e));
+
+        features = buildFeatureLookup(descNode, "Source_Device",
+                tag -> XMLConstants.FEATURE_BY_YNC_TAG.getOrDefault(tag, null),
+                (feature, e) -> new FeatureDescriptor(feature, e));
+
+        logger.debug("Found system {}, zones {}, features {}", system != null ? 1 : 0, zones.size(), features.size());
+    }
+
+    /**
+     * Tires to get the XML descriptor for the AVR
+     * @param con
+     * @return
+     */
+    private Node tryGetDescriptor(XMLConnection con) {
+        for (String path: Arrays.asList("/YamahaRemoteControl/desc.xml", "/YamahaRemoteControl/UnitDesc.xml")) {
+            try {
+                String descXml = con.getResponse(path);
+                Document doc = XMLUtils.xml(descXml);
+                Node root = doc.getFirstChild();
+                if (root != null && "Unit_Description".equals(root.getNodeName())) {
+                    logger.debug("Retrieved descriptor from {}", path);
+                    return root;
+                }
+                logger.debug("The {} response was invalid: {}", path, descXml);
+            } catch (IOException e) {
+                // The XML document under specified path does not exist for this model
+                logger.debug("No descriptor at path {}", path);
+            } catch (Exception e) {
```

Comment:
The point here is that 
a) attempt to see if any of the known feature descriptor endpoints exist (IOException) and 
b) the Exception is catched to ensure that feature detection does not break anything, for example what if on one Yamaha model the XML file was returned, but actually was not a parsable xml...

Reply:
> but actually was not a parsable xml

Sounds like a very specific exception

Reply:
Yes, for the parsing example provided, but extrapolating the weird things I saw I am accounting for much greater evil than just parsing :)

Reply:
Okay, next time please focus on limiting the exceptions you catch, because that will help in not hiding away programming errors in the code. 

I can image that it requires too much retesting for now.